code/espurna/sensors/ADE7953Sensor.h

```diff
@@ -0,0 +1,223 @@
+        String description() {
+            char buffer[25];
+            snprintf(buffer, sizeof(buffer), "ADE7953 @ I2C (0x%02X)", _address);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {            
+            int relay = index / ADE7953_TOTAL_DEVICES;
+            index = index - (relay * ADE7953_TOTAL_DEVICES);
+            if (index == 0) return MAGNITUDE_VOLTAGE;                        
+            if (index == 1) return MAGNITUDE_CURRENT;
+            if (index == 2) return MAGNITUDE_POWER_ACTIVE;
+            if (index == 3) return MAGNITUDE_ENERGY;            
+            return MAGNITUDE_NONE;
+        }
+
+        // Pre-read hook (usually to populate registers with up-to-date data)
+        void pre() {                        
+            uint32_t active_power1 = 0;
+            uint32_t active_power2 = 0;
+            uint32_t current_rms = 0;
+            uint32_t current_rms1 = 0;
+            uint32_t current_rms2 = 0;
+            uint32_t voltage_rms = 0;            
+
+            voltage_rms = Ade7953Read(0x31C);      // Both relays
+            current_rms1 = Ade7953Read(0x31B);     // Relay 1
+            if (current_rms1 < 2000) {             // No load threshold (20mA)
+                current_rms1 = 0;
+                active_power1 = 0;
+            } else {
+                active_power1 = (int32_t)Ade7953Read(0x313) * -1;  // Relay 1
+                active_power1 = (active_power1 > 0) ? active_power1 : 0;
+            }
+            current_rms2 = Ade7953Read(0x31A);     // Relay 2
+            if (current_rms2 < 2000) {             // No load threshold (20mA)
+                current_rms2 = 0;
+                active_power2 = 0;
+            } else {
+                active_power2 = (int32_t)Ade7953Read(0x312);  // Relay 2
+                active_power2 = (active_power2 > 0) ? active_power2 : 0;
+            }
+            _voltage = (float) voltage_rms / ADE7953_UREF;            
+            
+            writeFloat(
+                ADE7953_ALL_RELAYS, 
+                (float)(current_rms1 + current_rms2) / (ADE7953_IREF * 10), 
+                (float)(active_power1 + active_power2) / (ADE7953_PREF / 10)
+            );
+            writeFloat(
+                ADE7953_RELAY_1, 
+                (float) current_rms1 / (ADE7953_IREF * 10), 
+                (float) active_power1 / (ADE7953_PREF / 10)
+            );    
+            writeFloat(
+                ADE7953_RELAY_2, 
+                (float)current_rms2 / (ADE7953_IREF * 10), 
+                (float)active_power2 / (ADE7953_PREF / 10)
+            );
+        }
+
+        void writeFloat(unsigned int relay, float current, float power) {
+            float* current_reading = &_readings[relay].current;
```

Comment:
regarding exception, you sure readings array is initialized so [relay] access is usable? vector[] does not check bounds and neither is at(), since there are no exceptions.
btw, `auto& reading_ref = &_readings[relay]; reading_ref.current = current`? (may be fuzzy on the proper syntax)